File: db-service/src/services/pdf_label_service.js
/**
 * Génère une étiquette de maintenance préventive au format A6 paysage.
 * Utilise PDFKit pour la génération serveur-side.
 *
 * params = {
 *   equipmentName, serialNumber, department,
 *   technicianName, performedAt, nextPm, hospitalName
 * }
 */
async function generateMaintenanceLabelPdf(params) {
  // Import dynamique pour compatibilité ESM/CJS
  let PDFDocument;
  try {
    PDFDocument = require('pdfkit');
  } catch (_) {
    // pdfkit non installé — retourner un PDF minimal d'erreur
    return Buffer.from('%PDF-1.4\n1 0 obj\n<<\n/Type /Catalog\n/Pages 2 0 R\n>>\nendobj\n');
  }

  const {
    equipmentName = '',
    serialNumber  = null,
    department    = '',
    technicianName = '',
    performedAt   = '',
    nextPm        = '',
    hospitalName  = 'Hôpital de District de Kabutare',
  } = params;

  // A6 paysage : 419.53pt × 297.64pt (148mm × 105mm à 72dpi)
  const A6_W = 419.53;
  const A6_H = 297.64;

  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({
      size: [A6_W, A6_H],
      margin: 28,
      info: { Title: `Étiquette PM — ${equipmentName}` },
    });

    const buffers = [];
    doc.on('data', chunk => buffers.push(chunk));
    doc.on('end', () => resolve(Buffer.concat(buffers)));
    doc.on('error', reject);

    const marginX = 28;
    const contentW = A6_W - 2 * marginX;
    let y = 28;

    // ── En-tête hôpital ──────────────────────────────────────────────────────
    doc.rect(marginX, y, contentW, 24).fill('#1565c0');
    doc.fillColor('white').fontSize(10).font('Helvetica-Bold');
    doc.text(hospitalName, marginX + 4, y + 7, { width: contentW - 8, align: 'center' });

    y += 30;

    // ── Séparateur ───────────────────────────────────────────────────────────
    doc.moveTo(marginX, y).lineTo(marginX + contentW, y)
       .strokeColor('#9e9e9e').lineWidth(0.5).stroke();
    y += 8;

    // ── Lignes équipement ────────────────────────────────────────────────────
    doc.fillColor('#212121').fontSize(9).font('Helvetica');
    _row(doc, marginX, y, contentW, 'Équipement', equipmentName);
    y += 18;

    if (serialNumber) {
      _row(doc, marginX, y, contentW, 'N° série', serialNumber);
      y += 18;
    }

    _row(doc, marginX, y, contentW, 'Département', department);
    y += 18;

    // ── Séparateur ───────────────────────────────────────────────────────────
    doc.moveTo(marginX, y).lineTo(marginX + contentW, y)
       .strokeColor('#9e9e9e').lineWidth(0.5).stroke();
    y += 8;

    // ── Lignes maintenance ───────────────────────────────────────────────────
    _row(doc, marginX, y, contentW, 'Technicien', technicianName);
    y += 18;

    _row(doc, marginX, y, contentW, 'Date maint.', _formatDate(performedAt));
    y += 18;

    // Prochaine maintenance en bleu gras
    doc.font('Helvetica-Bold').fontSize(9).fillColor('#1565c0');
    _row(doc, marginX, y, contentW, 'Prochaine', _formatDate(nextPm), true);

    doc.end();
  });
}

// ── Helpers ──────────────────────────────────────────────────────────────────

function _row(doc, x, y, width, label, value, highlight = false) {
  const labelW = 80;
  doc.font('Helvetica').fontSize(9).fillColor('#757575');
  doc.text(label, x, y, { width: labelW });
  doc.font(highlight ? 'Helvetica-Bold' : 'Helvetica')
     .fontSize(9)
     .fillColor(highlight ? '#1565c0' : '#212121');
  doc.text(value || '—', x + labelW, y, { width: width - labelW });
}

function _formatDate(iso) {
  if (!iso || iso.length < 10) return iso || '—';
  const parts = iso.substring(0, 10).split('-');
  if (parts.length !== 3) return iso;
  return `${parts[2]}/${parts[1]}/${parts[0]}`;
}

module.exports = { generateMaintenanceLabelPdf };

File: flutter-app/lib/services/pdf_label_service.dart
import 'dart:typed_data';

import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;

/// Génère une étiquette de maintenance préventive au format A6 paysage.
///
/// Aucun appel réseau — toutes les données sont passées en paramètre
/// depuis [EquipmentMaintenanceTab] après validation PM.
class PdfLabelService {
  static const _primary   = PdfColors.blue800;
  static const _border    = PdfColors.grey400;
  static const _textMuted = PdfColors.grey600;

  static const _hospitalName = 'Hôpital de District de Kabutare';

  /// Construit et retourne les octets du PDF A6 paysage.
  static Future<Uint8List> generateMaintenanceLabel({
    required String equipmentName,
    required String? serialNumber,
    required String department,
    required String technicianName,
    required String performedAt,
    required String nextPm,
  }) async {
    final doc = pw.Document();

    // Format A6 paysage : 148mm × 105mm
    const pageFormat = PdfPageFormat(
      148 * PdfPageFormat.mm,
      105 * PdfPageFormat.mm,
    );

    final performedFormatted = _formatDate(performedAt);
    final nextFormatted      = _formatDate(nextPm);

    doc.addPage(
      pw.Page(
        pageFormat: pageFormat,
        margin: pw.EdgeInsets.all(10 * PdfPageFormat.mm),
        build: (pw.Context context) => pw.Column(
          crossAxisAlignment: pw.CrossAxisAlignment.stretch,
          children: [
            // ── En-tête hôpital ───────────────────────────────────────
            pw.Container(
              padding: const pw.EdgeInsets.symmetric(
                  horizontal: 8, vertical: 4),
              decoration: pw.BoxDecoration(
                color: _primary,
                borderRadius: pw.BorderRadius.circular(4),
              ),
              child: pw.Text(
                _hospitalName,
                style: pw.TextStyle(
                  fontSize: 10,
                  fontWeight: pw.FontWeight.bold,
                  color: PdfColors.white,
                ),
                textAlign: pw.TextAlign.center,
              ),
            ),
            pw.SizedBox(height: 6),

            // ── Séparateur ────────────────────────────────────────────
            pw.Divider(color: _border, thickness: 0.5),
            pw.SizedBox(height: 4),

            // ── Informations équipement ───────────────────────────────
            _labelRow('Équipement', equipmentName),
            if (serialNumber != null && serialNumber.isNotEmpty)
              _labelRow('N° série', serialNumber),
            _labelRow('Département', department),

            pw.SizedBox(height: 4),
            pw.Divider(color: _border, thickness: 0.5),
            pw.SizedBox(height: 4),

            // ── Informations maintenance ──────────────────────────────
            _labelRow('Technicien', technicianName),
            _labelRow('Date maint.', performedFormatted),
            _labelRow(
              'Prochaine',
              nextFormatted,
              valueStyle: pw.TextStyle(
                fontSize: 10,
                fontWeight: pw.FontWeight.bold,
                color: _primary,
              ),
            ),
          ],
        ),
      ),
    );

    return doc.save();
  }

  // ── Helpers ─────────────────────────────────────────────────────────────────

  static pw.Widget _labelRow(
    String label,
    String value, {
    pw.TextStyle? valueStyle,
  }) {
    return pw.Padding(
      padding: const pw.EdgeInsets.symmetric(vertical: 2),
      child: pw.Row(
        crossAxisAlignment: pw.CrossAxisAlignment.start,
        children: [
          pw.SizedBox(
            width: 65,
            child: pw.Text(
              label,
              style: pw.TextStyle(
                fontSize: 9,
                color: _textMuted,
              ),
            ),
          ),
          pw.Expanded(
            child: pw.Text(
              value,
              style: valueStyle ??
                  pw.TextStyle(
                    fontSize: 10,
                    fontWeight: pw.FontWeight.bold,
                  ),
            ),
          ),
        ],
      ),
    );
  }

  /// Formate un ISO datetime → DD/MM/YYYY
  static String _formatDate(String iso) {
    if (iso.length < 10) return iso;
    final s = iso.substring(0, 10).split('-');
    if (s.length != 3) return iso;
    return '${s[2]}/${s[1]}/${s[0]}';
  }
}
